src/rust/src/wallet.rs

```diff
@@ -135,33 +155,113 @@ impl Wallet {
+        // block must be checkpointed
+        if self
+            .last_checkpoint
+            .iter()
+            .any(|last_height| block_height != *last_height + 1)
+        {
+            return false;
+        }
+
         self.witness_tree.checkpoint();
+        self.last_checkpoint = Some(block_height);
+        true
     }
 
-    pub fn rewind_witness_tree(&mut self) -> bool {
-        self.witness_tree.rewind()
+    /// Returns whether or not a checkpoint has been created. If no checkpoint exists,
+    /// the wallet has not yet observed any blocks.
+    pub fn is_checkpointed(&self) -> bool {
+        self.last_checkpoint.is_some()
+    }
+
+    /// Rewinds the note commitment tree to the last checkpoint. If the rewind operation
+    /// succeeds, this also decrements the checkpoint height.
+    pub fn rewind(&mut self, to_height: BlockHeight) -> bool {
+        if let Some(checkpoint_height) = self.last_checkpoint.as_mut() {
+            // if the rewind length exceeds the number of checkpoints we have stored,
+            // return failure.
+            let mut blocks_to_rewind: i64 =
+                <i64>::from(to_height) - <i64>::from(*checkpoint_height);
+            if blocks_to_rewind > self.witness_tree.checkpoints().len().try_into().unwrap() {
+                return false;
+            }
+
+            // when we rewind, we need to remove the spentness marks for the notes we're tracking
+            while blocks_to_rewind > 0 {
+                if !self.witness_tree.rewind() {
+                    return false;
+                }
+
+                *checkpoint_height = *checkpoint_height - 1;
+                for (txid, n) in self.wallet_tx_notes.iter() {
+                    if n.tx_height > *checkpoint_height {
+                        self.spent_notes.retain(|o| o.txid != *txid)
+                    }
+                }
+                // remove any notes that were mined since the last checkpoint
+                self.wallet_tx_notes
+                    .retain(|_, n| n.tx_height <= *checkpoint_height);
+                // reset our last observed height to ensure that notes added in the future are
+                // from a new block
+                self.last_observed = Some(LastObserved {
+                    block_height: *checkpoint_height,
+                    block_tx_idx: None,
+                });
+
+                blocks_to_rewind -= 1;
+            }
+
+            true
+        } else {
+            false
+        }
     }
 
     /// Add note data for those notes that are decryptable with one of this wallet's
-    /// incoming viewing keys to the wallet, and return the indices of the actions
-    /// that we were able to decrypt.
+    /// incoming viewing keys to the wallet, and return the indices of the actions that we
+    /// were able to decrypt.
     pub fn add_notes_from_bundle(
         &mut self,
+        block_height: BlockHeight,
```

Comment:
We can't set this for transactions we detect via the mempool (and want to display in RPCs via `minconf=0`). It necessarily needs to be optional:
```suggestion
        block_height: Option<BlockHeight>,
```

Reply:
Instead of making it optional, I changed it so that TxNotes::tx_height only gets set when the witness tree is updated.